### ISSUE TITLE
testing strategy for web ui

### ISSUE THREAD

**Issue by jeffmcaffer** [clearlydefined/website#24]
Need to investigate the right/best way to test React components and apps, reducers, ...

* Pick a strategy
* implement example tests for each different kind of thing (component, container, reducer, ...)
* hook into build to ensure tests are run before deployment

**Comment by mynameistechno** [comment]
For reference https://facebook.github.io/jest/docs/en/tutorial-react.html

**Comment by mynameistechno** [comment]
For end-to-end testing we use nightwatchjs, but this still leverages selenium. Recently Chrome supports headless mode, and there are some newer browser automation tooling built around this that we should explore if we want to have browser-level/end-to-end type tests.

**Comment by jeffmcaffer** [comment]
Great. I am new to this space so any help here would be greatly appreciated. Would be awesome if someone picked and did some examples.

**Comment by AlexWebYourmind** [comment]
@mynameistechno  nightwatchjs (leveraging selenium) is a  great option indeed for E2E. 

As an alternative, may I also recommend to checkout cucumber (https://cucumber.io/) which has its own js implementation --> https://github.com/cucumber/cucumber-js and lets you write plain english tests (after creating different "grammars") and helps in adopting a BDD/TDD driven approach (but that's not a constraint).

Cucumber can of course leverage Selenium and I have been using that on many projects and always turned out quite useful. Also tests can be written resembling User Stories and Acceptance Criteria where needed ( I know we are not adopting SCRUM yet).

Long story short, we could start working towards introducing E2E Tests covering all the CD Use Cases (from a User/Actor perspective) and gradually achieve a good test coverage.

This is definitely doable with Nightwatch as well, it's just a matter of deciding which E2E tool we prefer.

**Comment by mynameistechno** [comment]
> @mynameistechno nightwatchjs (leveraging selenium) is a great option indeed for E2E.

Yes I was thinking the same. I.e. given the current lack of unit tests, writing some end-to-end tests for the happy paths would be the easiest way to add some sort of coverage to help prevent regressions. E2E tests shouldn't be used for code coverage however, since e2e tests don't have visibility into the code and are abstracted away from them. Having said that, they provide great functional coverage.

> As an alternative, may I also recommend to checkout cucumber (https://cucumber.io/) which has its own js implementation --> https://github.com/cucumber/cucumber-js

I've heard of cucumber before, I believe it's popular with the Ruby community. I'll take a look. Looks like people use it with protractor, which leverages selenium. We can go that route, but I was hoping to avoid selenium/java altogether. I've been meaning to play around with Chrome headless + Puppeteer + jest-puppeteer 
* https://pptr.dev/
* https://github.com/smooth-code/jest-puppeteer

I'll play around with this setup over the next week or so and report back.

**Comment by storrisi** [comment]
I've played a bit around tests, and as @mynameistechno already mentioned for e2e, an upgrade to react-scripts 2 is required.
I've made the upgrade on a local branch and it seems to have no regression, just only a couple of warnings to fix about links and input text.
I just want only to be sure about that there aren't any problems related to react-scripts before opening a PR.

**Comment by mynameistechno** [comment]
Thanks @storrisi. Create React App 2.0 (aka `react-scripts`) is out of beta and has officially been released, so I think it's fairly stable at this point. It was in beta for a long time as they were waiting for babel 7.0 stable to be released (which it was recently). I'll update this PR when I get a chance. 

The main thing left to sort out with e2e testing is the backend service piece. Unfortunately we haven't had time to work on that part of it due to some resourcing and scheduling conflicts.

**Comment by storrisi** [comment]
I've made some tests about react-scripts 2.0.0 and we had no regressions or conflicts. Good to go.
I'm starting building e2e tests based on puppeteer.
We can close this.

**Comment by jeffmcaffer** [comment]
Can you update https://github.com/clearlydefined/website/pull/286 for the basics and at least one example test that builds and runs? This issue ought not be closed until the changes are in.
